pennylane/devices/preprocess.py

```diff
@@ -0,0 +1,332 @@
+        it returns a QNode with the transform added to its transform program.
+        If a tape is passed, returns a tuple containing a list of
+        quantum tapes to be evaluated, and a function to be applied to these
+        tape executions.
+
+    Raises:
+        DecompositionUndefinedError: if an operator is not accepted and does not define a decomposition
+
+        DeviceError: If the decomposition enters and infinite loop and raises a ``RecursionError``.
+    """
+
+    # TODO: check if this pre-check actually leads to any performance improvements
+    if not all(stopping_condition(op) for op in tape.operations):
+        try:
+            # don't decompose initial operations if its StatePrepBase
+            prep_op = (
+                [tape[0]] if isinstance(tape[0], StatePrepBase) and skip_initial_state_prep else []
+            )
+
+            new_ops = [
+                final_op
+                for op in tape.operations[bool(prep_op) :]
+                for final_op in _operator_decomposition_gen(op, stopping_condition, name)
+            ]
+        except RecursionError as e:
+            raise DeviceError(
+                "Reached recursion limit trying to decompose operations. "
+                "Operator decomposition may have entered an infinite loop."
+            ) from e
+        tape = qml.tape.QuantumScript(prep_op + new_ops, tape.measurements, shots=tape.shots)
+
+    return (tape,), null_postprocessing
+
+
+@transform
+def validate_observables(
+    tape: qml.tape.QuantumTape,
+    stopping_condition: Callable[[qml.operation.Operator], bool],
+    name: str = "device",
+) -> (Sequence[qml.tape.QuantumTape], Callable):
+    """Validates the observables and measurements for a circuit.
+
+    Args:
+        tape (QuantumTape): a quantum circuit.
+        observable_stopping_condition (callable): a function that specifies whether or not an observable is accepted.
+        name (str): the name of the device to use in error messages.
+
+    Returns:
+        pennylane.QNode or qfunc or Tuple[List[.QuantumTape], Callable]: If a QNode is passed,
+        it returns a QNode with the transform added to its transform program.
+        If a tape is passed, returns a tuple containing a list of
+        quantum tapes to be evaluated, and a function to be applied to these
+        tape executions.
+
+    Raises:
+        DeviceError: if an observable is not supported
+
+    """
+    for m in tape.measurements:
+        if m.obs is not None:
+            if isinstance(m.obs, Tensor):
+                if any(not stopping_condition(o) for o in m.obs.obs):
+                    raise DeviceError(f"Observable {m.obs} not supported on {name}")
+            elif not stopping_condition(m.obs):
+                raise DeviceError(f"Observable {m.obs} not supported on {name}")
+
+    return (tape,), null_postprocessing
+
+
+@transform
+def validate_measurements(
```

Comment:
What is the difference between this and `validate_observables`?

Reply:
They're quite similar, and could potentially be combined. But we have essentially three "independent" checks with three independent keyword arguments.

So we could have one transform with three keyword arguments, three transforms with one keyword arguments, one a hybrid of the two (current version).